crates/bevy_ecs/src/schedule_v3/schedule.rs

```diff
@@ -0,0 +1,1027 @@
+
+        // check flattened dependencies for cycles
+        let flat_scc = tarjan_scc(&dependency_flattened);
+        if self.contains_cycles(&flat_scc) {
+            self.report_cycles(&flat_scc);
+            return Err(BuildError::DependencyCycle);
+        }
+
+        self.dependency_flattened.graph = dependency_flattened;
+        self.dependency_flattened.topsort =
+            flat_scc.into_iter().flatten().rev().collect::<Vec<_>>();
+
+        let flat_results = check_graph(
+            &self.dependency_flattened.graph,
+            &self.dependency_flattened.topsort,
+        );
+
+        // flatten allowed ambiguities
+        let mut ambiguous_with_flattened = UnGraphMap::new();
+        for (lhs, rhs, _) in self.ambiguous_with.all_edges() {
+            match (lhs, rhs) {
+                (NodeId::System(_), NodeId::System(_)) => {
+                    ambiguous_with_flattened.add_edge(lhs, rhs, ());
+                }
+                (NodeId::Set(_), NodeId::System(_)) => {
+                    for &lhs_ in &systems[&lhs] {
+                        ambiguous_with_flattened.add_edge(lhs_, rhs, ());
+                    }
+                }
+                (NodeId::System(_), NodeId::Set(_)) => {
+                    for &rhs_ in &systems[&rhs] {
+                        ambiguous_with_flattened.add_edge(lhs, rhs_, ());
+                    }
+                }
+                (NodeId::Set(_), NodeId::Set(_)) => {
+                    for &lhs_ in &systems[&lhs] {
+                        for &rhs_ in &systems[&rhs] {
+                            ambiguous_with_flattened.add_edge(lhs_, rhs_, ());
+                        }
+                    }
+                }
+            }
+        }
+
+        self.ambiguous_with_flattened = ambiguous_with_flattened;
+
+        // check for conflicts
+        let mut conflicting_systems = Vec::new();
+        for &(a, b) in flat_results.disconnected.iter() {
+            if self.ambiguous_with_flattened.contains_edge(a, b)
+                || self.ambiguous_with_all.contains(&a)
+                || self.ambiguous_with_all.contains(&b)
+            {
+                continue;
+            }
+
+            let system_a = self.systems.get(&a).unwrap();
+            let system_b = self.systems.get(&b).unwrap();
+            if system_a.is_exclusive() || system_b.is_exclusive() {
+                conflicting_systems.push((a, b, Vec::new()));
+            } else {
+                let access_a = system_a.component_access();
+                let access_b = system_b.component_access();
+                if !access_a.is_compatible(access_b) {
+                    let conflicts = access_a.get_conflicts(access_b);
+                    conflicting_systems.push((a, b, conflicts));
+                }
+            }
+        }
+
+        if self.contains_conflicts(&conflicting_systems) {
+            self.report_conflicts(&conflicting_systems);
+            return Err(BuildError::Ambiguity);
```

Comment:
Alright.

Reply:
When that future time does come I'd prefer to be able to opt-in to panic on ambiguity in certain restricted settings, it would save effort for anyone that has such a requirement from maintaining patches against upstream. This is just a vote to have a good talk about it, it's pretty much mandatory in some uses.